### DISCUSSION TITLE
Why is brightness changing when using the Malvar algorithm?

### DISCUSSION THREAD

**Discussion by song4me** [colour-science/colour-demosaicing#34] [Q&A]
Hi,
When using the Malvar algorithm I get an output image which is much brighter than the original input image.
This change is different for different images. What is the reason this happens?

**Comment by thomasmansencal** [comment]
Hi @song4me,

How are you reading the image? Is it a linear floating-point image in [0, 1] range?

Cheers,

Thomas

**Comment by song4me** [comment]
Hi Thomas,
Thanks for your quick reply.
I read the image using `colour.io.read_image` and indeed got a linear floating-point image in the range of [0, 1].
I disabled the `eotf_inverse_sRGB` function (which was causing some of the brightness change and is not needed for me), but there's still a slight change in colors which I can't trace.

The code I'm using is the one below

```
IMAGE = colour.io.read_image('in.bmp')
im_out = colour_demosaicing.demosaicing_CFA_Bayer_Malvar2004(IMAGE, pattern='RGGB')
colour.io.write_image(im_out, 'out.bmp')

```

Thanks,
Shirley

**Comment by thomasmansencal** [comment]
Hello,

Would you have an example and/or the source image?

Cheers,

Thokas

**Comment by song4me** [comment]
Hi,
I'm attaching the input :
![in](https://user-images.githubusercontent.com/22259271/125608279-ec463b0a-8d41-4a12-9e43-49534f9b430d.png)
Result with opencv (for color comparison)
![dbyr_opencv](https://user-images.githubusercontent.com/22259271/125608341-47c6f078-54ed-429b-ba24-5c51820d62e2.png)
Results with Malvar from the colour demosaicing lib: 
![dbyr_malvar](https://user-images.githubusercontent.com/22259271/125608304-303fe41f-0509-42b1-8356-29baa7ffb03f.png)
thanks

**Reply by thomasmansencal** [reply]
Hello @song4me,

Thanks this is helpful! Something that makes me suspicious immediately is that the CFA data seems to be non-linearly encoded, which is fine in itself, but it should be linearized before doing anything! The other suspicious aspect is that the colours on the colour checker look almost too good for a camera outputting data in its native space. Usually, they tend to be greenish because of increased green sensitivity and the red spectral sensitivity offset.

I did take the CFA data as is, and ran it through the various algorithms we implement:

![image](https://user-images.githubusercontent.com/99779/125704547-311fa5a9-8e25-4308-83f9-2b7acc69100b.png)
![image](https://user-images.githubusercontent.com/99779/125704558-bdc8e1b3-1319-45d6-80e1-c9da919456a8.png)
![image](https://user-images.githubusercontent.com/99779/125704566-6b99d757-7ed5-4e25-b259-9017395dbf03.png)

As you can see, abstracting differences in demosaicing algorithm, they are consistently the same and are close to the OpenCV "look".

In any case, I would try to figure out what is the decoding colour component transfer function (CCTF) for your camera and then demosaic the linearized version.

Cheers,

Thomas

**Comment by song4me** [comment]
Thanks,
I do work on gamma corrected images (power of ~(1/2.2) ) 
How come your results are different than mine? My Malvar results are a bit darker than yours

I tested on linear images as well and get a similar behavior

**Reply by thomasmansencal** [reply]
I would say that the image is not decoded properly. How are you reading the image?

**Reply by song4me** [reply]
Like I write above, 
```
IMAGE = colour.io.read_image('in.bmp')
im_out = colour_demosaicing.demosaicing_CFA_Bayer_Malvar2004(IMAGE, pattern='RGGB')
colour.io.write_image(im_out, 'out.bmp')
```

**Reply by thomasmansencal** [reply]
I'm wondering if the BMP decoder of ImageIO (PIL by default?) is not doing something funny here. Is the BMP image here a single channel image?

**Reply by song4me** [reply]
Yes, it is

**Reply by thomasmansencal** [reply]
Ok, I'm assuming that this is the issue then, it is not being read/decoded properly, the PNG variant was though.

**Reply by song4me** [reply]
Your question led me to the right direction - when saving with opencv's imwrite instead if the image io the colors are written correctly.
Everything works great now, thanks again for the assistance!

**Reply by thomasmansencal** [reply]
Great to hear!
